eth/protocols/snap/sync.go

```diff
@@ -2121,6 +2162,50 @@ func (s *Syncer) processTrienodeHealResponse(res *trienodeHealResponse) {
 		log.Crit("Failed to persist healing data", "err", err)
 	}
 	log.Debug("Persisted set of healing data", "type", "trienodes", "bytes", common.StorageSize(batch.ValueSize()))
+
+	// Calculate the processing rate of one filled trie node
+	rate := float64(fills) / (float64(time.Since(start)) / float64(time.Second))
+
+	// Update the currently measured trienode queueing and processing throughput.
+	//
+	// The processing rate needs to be updated uniformly independent if we've
+	// processed 1x100 trie nodes or 100x1 to keep the rate consistent even in
+	// the face of varying network packets. As such, we cannot just measure the
+	// time it took to process N trie nodes and update once, we need one update
+	// per trie node.
+	//
+	// Naively, that would be:
+	//
+	//   for i:=0; i<fills; i++ {
+	//     healRate = (1-measurementImpact)*oldRate + measurementImpact*newRate
+	//   }
+	//
+	// Essentially, a recursive expansion of HR = (1-MI)*HR + MI*NR.
+	//
+	// We can expand that forumal for the Nth item as:
+	//   HR(N) = (1-MI)^N*OR + (1-MI)^(N-1)*MI*NR + (1-MI)^(N-2)*MI*NR + ... + (1-MI)^0*MI*NR
+	//
+	// The above sequence is a geometric mean that can be summed to:
+	//   HR(N) = (1-MI)^N*(OR-NR) + NR
+	s.trienodeHealRate = gomath.Pow(1-trienodeHealRateMeasurementImpact, float64(fills))*(s.trienodeHealRate-rate) + rate
+
+	pending := atomic.LoadUint64(&s.trienodeHealPend)
+	if time.Since(s.trienodeHealThrottled) > time.Second {
+		// Periodically adjust the trie node throttler
+		if float64(pending) > 2*s.trienodeHealRate {
```

Comment:
This seems a bit odd to me. Basically, every second, you are going to either increase it by `33%`, or decrease it by `25%`. It will never just stay where it is, always make a pretty large jump every second.

Reply:
An alternative would be to have an interval where it stays the same instead of jumping; or a second alternative would be to have a multiplier proportional to the deviation from the ideal.

The problem is that the rate is very very unpredictable. Full nodes will probably incur a large load - since they have 16 children to check - whereas short nodes (with a leaf embedded) will have 0 load. So depending on what's delivered, the rate will go weird. I guess if we returned some metadata post processing about the "incurred" load, we might have a stabler rate, and thus wouldn't need to very agressively adjust the multipliers.

Reply:
Though in that case I would also need to know what kind I'm requesting, which gets nasty again. I think we should get the LRU implemented and see how that helps, if that gets rid of many of the reads, it might be less of an assymetry between the two types of nodes